src/node/routes/proxy.ts

```diff
@@ -0,0 +1,43 @@
+import { Request, Router } from "express"
+import qs from "qs"
+import { HttpCode, HttpError } from "../../common/http"
+import { authenticated, redirect } from "../http"
+import { proxy } from "../proxy"
+
+export const router = Router()
+
+const getProxyTarget = (req: Request, rewrite: boolean): string => {
+  if (rewrite) {
+    const query = qs.stringify(req.query)
+    return `http://127.0.0.1:${req.params.port}/${req.params[0] || ""}${query ? `?${query}` : ""}`
+  }
+  return `http://127.0.0.1:${req.params.port}/${req.originalUrl}`
+}
+
+router.all("/(:port)(/*)?", (req, res) => {
+  if (!authenticated(req)) {
+    // If visiting the root (/proxy/:port and nothing else) redirect to the
+    // login page.
+    if (!req.params[0] || req.params[0] === "/") {
+      return redirect(req, res, "login", {
+        to: `${req.baseUrl}${req.path}` || "/",
+      })
+    }
+    throw new HttpError("Unauthorized", HttpCode.Unauthorized)
+  }
+
+  // Absolute redirects need to be based on the subpath when rewriting.
+  ;(req as any).base = `${req.baseUrl}/${req.params.port}`
+
+  proxy.web(req, res, {
+    ignorePath: true,
+    target: getProxyTarget(req, true),
+  })
+})
+
+router.ws("/(:port)(/*)?", (socket, head, req) => {
```

Comment:
I could be totally wrong here, but I thought that Express ran handlers in order. So shouldn't `all` be hitting first here? Or does `ws` somehow pick up things that `all` is missing?

Reply:
With my patched router if the request is for a websocket then **only** the `ws` methods will be called. If it's not a websocket request then the `ws` methods will **never** be called.

Reply:
Was wrong about this, would need to patch `all` as well to fix this.